### ISSUE TITLE
z/OS builds broken: `fatal error: 'stdatomic.h' file not found`

### ISSUE THREAD

**Issue by richardlau** [libuv/libuv#3714]
* **Version**: v1.x
* **Platform**: z/OS

It looks like https://github.com/libuv/libuv/pull/3688 has broken the z/OS CI. e.g. https://ci.nodejs.org/view/libuv/job/libuv-test-commit-zos-cmake/1349/BUILDTYPE=Release,nodes=zos-s390x/console
```
/u/iojs/build/workspace/libuv-test-commit-zos-cmake/BUILDTYPE/Release/nodes/zos-s390x/src/unix/async.c:29:10: fatal error: 'stdatomic.h' file not found
#include <stdatomic.h>
         ^~~~~~~~~~~~~
8 warnings and 1 error generated.
Error while processing /u/iojs/build/workspace/libuv-test-commit-zos-cmake/BUILDTYPE/Release/nodes/zos-s390x/src/unix/async.c.
CCN0793(I) Compilation failed for file /u/iojs/build/workspace/libuv-test-commit-zos-cmake/BUILDTYPE/Release/nodes/zos-s390x/src/unix/async.c.  Object file not created.
CMakeFiles/uv_a.dir/build.make:326: recipe for target 'CMakeFiles/uv_a.dir/src/unix/async.c.o' failed
make: *** [CMakeFiles/uv_a.dir/src/unix/async.c.o] Error 12
```

cc @libuv/zos

**Comment by vtjnash** [comment]
Docs say it should be there (https://www.ibm.com/docs/en/zos/2.1.0?topic=c11-atomic-integer-types), but we might to make sure we are passing the right compiler flags (such as `-std=c11`)

**Comment by IgorTodorovskiIBM** [comment]
That header was not implemented in xlclang. That's one of the reasons we had assembly code to implement the atomic functions. We may want to consider moving up to the Clang based compiler (https://www-40.ibm.com/servers/resourcelink/svc00100.nsf/pages/openXlCC++11ForZOs24AndZOs25?OpenDocument) which has stdatomics.h.

**Comment by twose** [comment]
I am confused, we have `set(CMAKE_C_STANDARD 90)` in CMakeLIsts.txt but we are using c11 atomic at the same time... Does this mean we need to switch to the C11 standard now?

**Comment by richardlau** [comment]
This isn't just affecting z/OS.

https://ci.nodejs.org/view/libuv/job/libuv-test-commit/2259/
- All centos7/rhel7 builds are broken as they're using the default gcc 4.8.5 compiler.
- The AIX build using the xlc compiler.

**Comment by santigimeno** [comment]
I've opened https://github.com/nodejs/build/issues/3074 for the centos7 build. If it can't be done, should we remove those nodes?

**Comment by oerdnj** [comment]
@bnoordhuis While I agree that requiring C11 and stdatomic is fine.  It has been 11 years since the standard, we do have compatibility shim in BIND 9 to use old GCC builtins. I would be happy to contribute this to libuv.

**Comment by oerdnj** [comment]
Feel free to pull the contents of https://gitlab.isc.org/isc-projects/bind9/-/blob/main/lib/isc/include/isc/stdatomic.h

and re-license under libuv license. You have my permission.

**Comment by oerdnj** [comment]
Then you need something like this:

https://gitlab.isc.org/isc-projects/bind9/-/blob/main/lib/isc/include/isc/atomic.h#L16

to decide whether you want C11 implementation or the shim.

**Comment by oerdnj** [comment]
> The c11 atomic changes on seem to reduce compatibility of libuv, I know that neither node.js nor Julia is bothered by this, but it seems less friendly to other projects?

It’s the other way around - enforcing the maintainers to use outdated version of C is less friendly on the downstream project. C11/C17 simplifies the code, makes it easier to maintain.

As a side note, the gcc 4.8.5 has most of the bits in place, so RHEL/CentOS 7 is mostly fine with few shims like the one I’ve just posted above.

**Comment by vtjnash** [comment]
I don't think that code belongs in libuv, but it would make sense for someone that wants to maintain an old compiler toolchain to add that file/folder to their includes path when configuring their build.

**Comment by oerdnj** [comment]
That’s up to the libuv maintainers, I am merely offering help. For other people this is MPL 2.0 licensed, so you can pull the up-to-date version of the shim header from BIND 9.

**Comment by bnoordhuis** [comment]
Speaking for myself, I'd be okay with shimming as long as it doesn't add a lot of conditional logic to the build scripts or the source files.

**Comment by hswong3i** [comment]
My tricks (see  https://github.com/alvistack/libuv-libuv/commit/08a2d1733546544e097228e396cbe2780307c040#diff-e1ff52f666e1fea28db016fcbb281f953f8d2836593f6e99195c8898f26a65aeR27-R32): 
```
%if 0%{?rhel} == 7
BuildRequires: devtoolset-11
BuildRequires: devtoolset-11-gcc
BuildRequires: devtoolset-11-gcc-c++
BuildRequires: devtoolset-11-libatomic-devel
%endif

%build
%if 0%{?rhel} == 7
. /opt/rh/devtoolset-11/enable
%endif
%configure
%make_build
```
